include/ismrmrd/serialization.h

```diff
@@ -0,0 +1,161 @@
+#ifndef ISMRMRDSERIALIZATION_H
+#define ISMRMRDSERIALIZATION_H
+
+#include <cstdint>
+#include <exception>
+#include <iostream>
+#include <memory>
+
+#include "ismrmrd/ismrmrd.h"
+#include "ismrmrd/waveform.h"
+#include "ismrmrd/xml.h"
+
+/**
+ * @file serialization.h
+ *
+ * @brief Serialization functions for ISMRMRD data structures
+ *
+ * This file contains functions for serializing and deserializing ISMRMRD data
+ * The standalone function that serialize to and deserialize from a stream will
+ * serrialize the data structures without adding "message id" identifiers.
+ * The ProtocolSerializer and ProtocolDeserializer classes are used to to create streams
+ * that include the message id in front of each message.
+ *
+ */
+
+namespace ISMRMRD {
+
+const uint16_t ISMRMRD_MESSAGE_UNPEEKED = 0;
+const uint16_t ISMRMRD_MESSAGE_HEADER = 3;
+const uint16_t ISMRMRD_MESSAGE_CLOSE = 4;
+const uint16_t ISMRMRD_MESSAGE_ACQUISITION = 1008;
+const uint16_t ISMRMRD_MESSAGE_IMAGE = 1022;
+const uint16_t ISMRMRD_MESSAGE_WAVEFORM = 1026;
+
+// A wrapper, which we can use wrap any object that has a read(...) method
+class ReadableStream {
+public:
+    void read(char *buffer, size_t count) {
+        reader_impl->read(buffer, count);
+    }
+
+    template <class HAS_READ>
+    ReadableStream(HAS_READ &has_read) {
```

Comment:
Did you consider making these `shared_ptr`s intead of references?

Reply:
You could have both:
```c++
  template <class HAS_WRITE>
  WritableStream(std::shared_ptr<HAS_WRITE> has_write)
      : writer_impl_(new WritableStreamT<HAS_WRITE>(has_write)) {
  }

  template <class HAS_WRITE>
  WritableStream(HAS_WRITE& has_write)
      : writer_impl_(new WritableStreamT<HAS_WRITE>(std::shared_ptr<HAS_WRITE>(&has_write, [](HAS_WRITE*) {}))) {
  }
```

Reply:
shared_ptr would not work with C++98. 
It would also make [ownership somewhat unclear](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers) - but maybe we should make it clear that ReadableStream is a view? ReadableStreamView maybe?

Either way, in general, a reference is all we can ask for. Many I/O resources are not copyable and sometimes not even movable. 

My thought was actually that we would always rely on the implicit conversion to ReadableStream, and just call the functions with the raw types. I know there is an overhead in creating the unique_ptr inside, but it should be very small.

Reply:
If we want to target C++98 then you can forget everything I said. 

My original thinking was to prevent someone from writing code like this (which will result in bad things):
``` cpp
std::unique_ptr<WritableStream> CreateWritableStream() {
    std::ofstream ofs("foo.bin", std::ios::binary);
    return std::make_unique<WritableStream>(ofs);
}
```

By instead requiring them to do something like
``` cpp
std::unique_ptr<WritableStream> CreateWritableStream2() {
    auto ofs = std::make_shared<std::ofstream>("foo.bin", std::ios::binary);
    return std::make_unique<WritableStream>(ofs);
}
```

But of course this a shared ptr can also result in bad things when it attempts to destruct:
``` cpp
std::unique_ptr<WritableStream> CreateWritableStream3() {
    auto ofs = std::shared_ptr<std::ostream>(&std::cout);
    return std::make_unique<WritableStream>(ofs);
}
```
And the proper way to wrap cout in a shared_ptr is pretty cumbersome:
``` cpp
std::unique_ptr<WritableStream> CreateWritableStream3() {
    auto ofs = std::shared_ptr<std::ostream>(&std::cout, [](std::ostream*){});
    return std::make_unique<WritableStream>(ofs);
}
```

I like your WritableStreamView suggestion. I think that helps make the lack of ownership clear.

I also think being able to rely on implicit conversion seems like a neat idea.

Reply:
Quick question here. If we are truly targeting C++98, we probably have to get rid of the unique_ptr too and just have a `ReadableStreamView*` member that we manage ourselves?

Reply:
You're right. It is rather annoying to have to do that ourselves. 
We should perhaps discuss where our API boundary should go. After all, we're not implementing this feature for the C API, so we could make the streaming interface here C++11 only. @kspaceKelvin any thoughts?

What we **need** is to be C++98 compatible for the basic types and the xml-header. I'm working on how we might do that without breaking API compatibillity.

Reply:
I have made this C++98 compatible for now. It is one point to manage in this case.